sdk/core/azure-core-experimental/src/main/java/com/azure/core/experimental/credential/AccessTokenCache.java

```diff
@@ -0,0 +1,165 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.experimental.credential;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.MonoProcessor;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A token cache that supports caching a token and refreshing it.
+ */
+public class AccessTokenCache {
+    // The delay after a refresh to attempt another token refresh
+    private static final Duration REFRESH_DELAY = Duration.ofSeconds(30);
+    // the offset before token expiry to attempt proactive token refresh
+    private static final Duration REFRESH_OFFSET = Duration.ofMinutes(5);
+    private final AtomicReference<MonoProcessor<AccessToken>> wip;
+    private volatile AccessToken cache;
+    private volatile OffsetDateTime nextTokenRefresh = OffsetDateTime.now();
+    private final Supplier<Mono<AccessToken>> tokenSupplier;
+    private final Predicate<AccessToken> shouldRefresh;
+    private final ClientLogger logger = new ClientLogger(AccessTokenCache.class);
+
+    /**
+     * Creates an instance of RefreshableTokenCredential with default scheme "Bearer".
+     *
+     * @param tokenSupplier a method to get a new token
+     */
+    public AccessTokenCache(Supplier<Mono<AccessToken>> tokenSupplier) {
+        this.wip = new AtomicReference<>();
+        this.tokenSupplier = tokenSupplier;
+        this.shouldRefresh = accessToken -> OffsetDateTime.now()
+            .isAfter(accessToken.getExpiresAt().minus(REFRESH_OFFSET));
+    }
+
+    /**
+     * Asynchronously get a token from either the cache or replenish the cache with a new token.
+     * @return a Publisher that emits an AccessToken
+     */
+    public Mono<AccessToken> getToken() {
+        return getToken(this.tokenSupplier, false);
+    }
+
+    /**
+     * Asynchronously get a token from either the cache or replenish the cache with a new token.
+     *
+     * @param tokenSupplier The method to get a new token
+     * @param forceRefresh The flag indicating if the cache needs to be skipped and a token needs to be fetched via the
+     * credential.
+     * @return The Publisher that emits an AccessToken
+     */
+    public Mono<AccessToken> getToken(Supplier<Mono<AccessToken>> tokenSupplier, boolean forceRefresh) {
```

Comment:
I think the parameter name `forceRefresh` might be a bit misleading here since the Supplier might or might not return a new token. Customers have asked for APIs such as this looking for a way to guarantee they get a new token, and I worry that they might think passing true to `forceRefresh` will accomplish that when really it just invalidates our top level cache and they might still get a cached token from the underlying layers, such as MSAL, MSI Endpoint, or STS. I don't have a better name in mind, and we can ship the preview with this name if needed, but we should consider this before GA.

Reply:
we could consider names like, skipCache, fetchToken for the boolean variable in next release.